### PR TITLE
ConfStore Phase1

### PR DESCRIPTION
**Pull request by suryakumar1024**
Includes :
ConfStore class
Conf Class
Test cases for both ConfStore and Conf classes

### CODE REVIEW

File: py-utils/src/utils/conf_store/conf_cache.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+
+    def _get(self, data: dict, key: str):
+        """ Obtain value for the given key """
+        k = key.split('.', 1)
+        if k[0] not in data.keys(): return None
+        return self._get(data[k[0]], k[1]) if len(k) > 1 else data[k[0]]
+
+    def get(self, key: str = None):
+        """ Returns the value corresponding to the key """
+        if key is None:
```
[the remark this answers is not in the document]
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cache.py]: Fixed in current patch. That If key is None: condition is not required here as i already handled it in the same get method itself.

**Review by ujjwalpl — APPROVED** [review]

**Review by ajaysrivas — APPROVED** [review]

File: py-utils/src/utils/conf_store/conf_store.py [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+                src_index)
+
+        if dst_index not in self._cache.keys():
+            raise ConfStoreError(errno.EINVAL, "config index %s is not loaded",
+                dst_index)
+
+        i_src = iter(self._cache[src_index])
+        while True:
+            try:
+                key = next(i_src)
```
Comment (ujjwalpl) [inline py-utils/src/utils/conf_store/conf_store.py:133]: Please remove this and change to the logic as per the patch I sent.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/src/utils/conf_store/conf_store.py (hunk 2 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_store.py]: Updated in current patch.

File: py-utils/src/utils/conf_store/conf_cache.py [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+        k = key.split('.', 1)
+        if k[0] not in data.keys(): return 
+        if len(k) > 1: return self._delete(data[k[0]], k[1])
+        del data[k[0]]
+        if key in self._keys:
+            self._keys.remove(key)
+
+    def delete(self, key: str):
+        """ Delets a given key from the config """
+        self._delete(self._data, key)
```
Comment (sachinpunadikar) [inline py-utils/src/utils/conf_store/conf_cache.py:114]: After deleting the key, shouldn't we mark as dirty (like the way it has been done for set operation) ?

File: py-utils/src/utils/conf_store/conf_store.py [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+        """ Saves the given index configuration onto KV Store """
+        if index not in self._cache.keys():
+            raise ConfStoreError(errno.EINVAL, "config index %s is not loaded",
+                index)
+
+        self._cache[index].dump()
+
+    def get(self, index: str, key: str, default_val=None):
+        """
+        Paraeters:
```
Comment (sachinpunadikar) [inline py-utils/src/utils/conf_store/conf_store.py:60]: Add comment here (which was wrongly mentioned for other method):
Obtain value for the key from given configuration

File: py-utils/src/utils/conf_store/conf_store.py [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+                                 index)
+        if key is None:
+            raise ConfStoreError(errno.EINVAL, "can't able to find config key "
+                                               "%s in loaded config", key)
+        val = self._cache[index].get(key)
+        return default_val if val is None else val
+
+    def set(self, index: str, key: str, val):
+        """
+        Obtain value for the given configuration
```
Comment (sachinpunadikar) [inline py-utils/src/utils/conf_store/conf_store.py:82]: Incorrect comment, we are setting value here, not obtaining

**Review by sachinpunadikar — CHANGES_REQUESTED** [review]
Few comments, address them as appropriate

File: py-utils/src/utils/conf_store/conf_store.py (hunk 5 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_store.py]: Updated comment in current patch

File: py-utils/src/utils/conf_store/conf_store.py (hunk 4 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_store.py:62]: Updated comment in current patch

File: py-utils/src/utils/conf_store/conf_cache.py (hunk 3 above)
Comment (suryakumar1024) [reply py-utils/src/utils/conf_store/conf_cache.py:117]: Updated and added this logic in current patch

**Review by sachinpunadikar — APPROVED** [review]
Looks good to me

File: py-utils/setup.py [hunk 6]
```diff
@@ -56,6 +56,8 @@ def get_install_requirements() -> List:
                 'cortx.utils.ha',
                 'cortx.utils.validator',
                 'cortx.utils.kvstore',
+                'cortx.utils.test_kv_store',
```
Comment (ujjwalpl) [inline py-utils/setup.py:59]: You dont need to export test_kv_store. test should not even be part of rpm. This is why its outside of src.

**Review by ujjwalpl — CHANGES_REQUESTED** [review]

File: py-utils/setup.py (hunk 6 above)
Comment (suryakumar1024) [reply py-utils/setup.py]: Updated in current patch @ujjwalpl

**Review by ujjwalpl — APPROVED** [review]

**Pull request merged**